# Reproducible codes for ONTraC characterizes spatially continuous variations of tissue microenvironment through niche trajectory analysis

## Abstract

Recent technological advances enable mapping of tissue spatial organization at single‐ cell resolution, but methods for analyzing spatially continuous microenvironments are still lacking. We introduce ONTraC, a graph neural network‐based framework for con‐ structing spatial trajectories at niche‐level. Through benchmarking analyses using mul‐ tiple simulated and real datasets, we show that ONTraC outperforms existing methods. ONTraC captures both normal anatomical structures and disease‐associated tissue microenvironment changes. In addition, it identifies tissue microenvironment‐depend‐ ent shifts in gene expression, regulatory network, and cell–cell interaction patterns. Taken together, ONTraC provides a useful framework for characterizing the structural and functional organization of tissue microenvironments.

## Citation

**Wang, W.\*, Zheng, S.\*, Shin, C. S., Chávez-Fuentes J. C.  & [Yuan, G. C.](https://labs.icahn.mssm.edu/yuanlab/)$**. [ONTraC characterizes spatially continuous variations of tissue microenvironment through niche trajectory analysis](https://doi.org/10.1186/s13059-025-03588-5). *Genome Biol*, 2025.

## Other links

- [ONTraC GitHub](https://github.com/gyuanlab/ONTraC)
- [ONTraC PyPI](https://pypi.org/project/ONTraC/)
- [ONTraC Zenodo](https://doi.org/10.5281/zenodo.14171604)
- [ONTraC wesite](https://ontrac-website.readthedocs.io/en/latest/)
- [IO files for reproducing](https://doi.org/10.5281/zenodo.11186619)
