# Reproducible codes for Characterizing Spatially Continuous Variations of Tissue Microenvironment through Niche Trajectory Analysis

## Abstract

Recent technological advances enable mapping of tissue spatial organization at single-cell resolution, but methods for analyzing spatially continuous microenvironments are still lacking. We introduce ONTraC, a graph neural network-based framework for constructing spatial trajectories at niche-level. Through benchmarking analyses using multiple simulated and real datasets, we show that ONTraC outperforms existing methods. ONTraC captures both normal anatomical structures and disease-associated tissue microenvironment changes. In addition, it identifies tissue microenvironment-dependent shifts in gene expression, regulatory network, and cell-cell interaction patterns. Taken together, ONTraC provides a useful framework for characterizing the structural and functional organization of tissue microenvironments.

## Citation

Wen Wang1,\*, Shiwei Zheng1,\*, Sujung Crystal Shin1, Joselyn Cristina Ch√°vez-Fuentes1, Guo-Cheng Yuan1,$

1. Department of Genetics and Genomic Sciences, Icahn School of Medicine at Mount Sinai, New York, NY, 10029, USA

## Other links

- [ONTraC GitHub](https://github.com/gyuanlab/ONTraC)
- [ONTraC PyPI](https://pypi.org/project/ONTraC/)
- [ONTraC Zenodo](https://doi.org/10.5281/zenodo.14171604)
- [ONTraC wesite](https://ontrac-website-template.readthedocs.io/)
- [IO files for reproducing](https://doi.org/10.5281/zenodo.11186619)
